posts/2017-12-20-who-implements-typeclass.md

```diff
@@ -0,0 +1,565 @@
+---
+layout: post
+title: Who implements the typeclass instance?
+
+meta:
+  nav: blog
+  author: S11001001
+  pygments: true
+---
+
+The typeclass pattern in Scala invites you to place
+implementation-specific knowledge directly in the typeclass instances,
+with the interface defined as the typeclass’s abstract interface.
+
+However, GADTs permit a different organization of code. It is even
+possible to define a typeclass that seems to do nothing at all, yet
+still permits full type-safe typeclass usage.
+
+The possibilities between these two extremes form a design space. If
+you wish to practice ad-hoc polymorphism in Scala, this space is well
+worth exploring.
+
+## A glorified overloader
+
+Refactoring a set of overloads into a typeclass is a fine way to get
+some free flexibility and dynamism, because expressing overloads as a
+typeclass gives you free fixes for common overload problems.
+
+1. Methods calling the overloaded method do not themselves need to be
+   overloaded just to avoid suppressing the flexibility of the
+   overload beneath.
+1. Return-type overloading works, even in Scala, whose “normal”
```

Comment:
Nitpick: overloading by return type in Scala is done by some APIs, in particular spray, where it's called [the Magnet Pattern](http://spray.io/blog/2012-12-13-the-magnet-pattern/).

Reply:
272104c should help a little here too; thanks. (I mean to compare only to classic Java-style overloads, the sort you can do return-type overloading with in C++.)